### PR TITLE
feat(openclaw): survive the 2026.8.1 host cutover

### PR DESCRIPTION
**Pull request by Amyssjj**
Prepares the digital-me overlay for openclaw **2026.8.1** (npm `latest`). Grounded on a diff of the 2026.6.11 and 2026.8.1 sources, not on release notes.

There is no "OpenClaw 2.0" — upstream is calendar-versioned (npm carries only `0` and `2026` major prefixes). The real jump is `2026.6.11 → 2026.8.1`.

### 1. Conversation-hook gate (silent)
openclaw drops a "conversation" typed hook at registration when a **non-bundled** plugin has not set `hooks.allowConversationAccess: true` — a warn diagnostic and an early `return` (`plugins/registry-registrars-tools-hooks.ts:380`). The plugin still loads, still registers tools, still reports `Status: loaded`. Only the hooks are gone.

2026.8.1 adds `before_prompt_build` to `CONVERSATION_HOOK_NAMES` (`plugins/hook-types.ts:229`) — that is the entire wiki-recall path. Every digital-me plugin is non-bundled (`openclaw plugins inspect` → `Origin: global`).

**This is already biting today.** `agent_end` has been in the gated set since before 2026.6.11, so the M1 application-ack emitter is dark right now:

```
runtime=openclaw  knowledge_surfaced  505  last 2026-08-31 22:30   <- before_prompt_build, working
runtime=openclaw  assistant_ack       134  last 2026-08-24 22:30   <- agent_end, gated
```

The installer now writes the grant. **Only `digital-me-recall`** — `digital-me-brain` registers tools and zero hooks, so granting it would hand out a capability it never exercises (least privilege).

### 2. Memory-search namespace cutover (fatal)
2026.7.1 moved the block from `agents.defaults.memorySearch` to root `memory.search`. Both schemas are `.strict()` and the config loader **throws** on an unknown key (`config/io.load.ts:127` → `throwInvalidConfig`).

So this is not "corpus silently unindexed" — **the gateway will not start**. And since 6.11's root `MemorySchema` is `{backend, citations, qmd}` with no `search`, while 8.1's `agents.defaults` has no `memorySearch`, **no single config is valid on both versions**. The migration must happen inside the binary-swap window.

The installer now detects the running host — source checkout `package.json` → `openclaw --version` → the config's own `meta.lastTouchedVersion` — writes the layout that host accepts, and migrates an existing block **in either direction** (forward on update, back on rollback). `doctor` reads whichever layout is present and names that key in its repair hint.

All six sub-keys we write or preserve (`enabled`, `extraPaths`, `provider`, `fallback`, `model`, `remote`) exist in both versions' `MemorySearchSchema`, so the move is lossless.

### 3. Nothing could have caught either one
- The updater's overlay smoke is `node --check` — **syntax only**. It cannot load the bundle because `openclaw/*` is externalized, so an SDK export removal would also pass it.
- The post-update sweep covered exactly one pipeline (`digest-smoke`).
- No CI job references openclaw at all, despite `docs/UPSTREAM-ADAPTATION-CONSTRAINT.md` prescribing exactly that canary.
- `deploy.ts`'s `parseRecallAckMode` greps for `assistant_ack=…`, but that value was a **hardcoded string literal** — it fingerprinted the build, not the registration.

Adds `scripts/verify_openclaw_hooks.sh` to the update profile. It fails on a missing grant, on a host block diagnostic in the gateway log, or on the plugin's own boot self-check reporting `conversation_hooks=BLOCKED`. Verified red on the current unfixed live config, before any fix was applied.

The recall registration line is now **derived** from the hooks the host will honour, and a missing grant logs a loud warning at boot — so deploy's existing fingerprint parser becomes a real canary for free.

## Also
`MAX_TESTED_OPENCLAW_VERSION` 2026.6.10 → 2026.8.1. It was *below* the installed 2026.6.11, so the "untested host" warning fired on every boot and had stopped being a signal.

## Not in this PR
Node 24 (openclaw ≥2026.7.1 requires `>=22.22.3 <23 || >=24.15.0 <25 || >=25.9.0`; the host runs 25.4.0, which satisfies none — `entry.ts:140` `assertSupportedRuntime()` hard-exits) and the live update itself. Both are host operations, tracked separately.

## Test plan
- `pnpm run ci` green — sanitize, build, typecheck, lint, knip, **100% coverage workspace-wide**
- New coverage: layout detect/resolve/migrate in both directions, grant merge (incl. never overriding a deliberate `false`), host-version detection precedence, calendar-version comparison (`2026.10.1 > 2026.9.1` is not string order), and the hook-grant doctor check
- Memory tests made **hermetic** — they previously read the host's real `openclaw --version` to pick a layout, so they would have behaved differently in CI

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by Amyssjj** [comment]
### Added: openclaw memory-index leak detection (`03bb73c`)

While verifying the 2026.8.1 upgrade live, a second silent openclaw failure surfaced — same shape as the hook gate, different mechanism.

openclaw's full reindex builds into `<agent>.sqlite.memory-reindex-<uuid>` and **does not delete it when the build aborts** on a revision race. Each failed attempt strands a full copy of the index. Observed in one afternoon:

| agent | orphans | size |
|---|---|---|
| coo | 4 | 4.8 GiB |
| main | 1 | 305 MB |

Every other signal stayed green throughout: `openclaw memory status` reported `Dirty: no` with 1830/1830 files indexed, search kept returning hits, and the gateway probe passed. The only symptoms were disk quietly disappearing and RSS spiking to ~1.9 GiB per doomed rebuild.

This is an **upstream defect** in `memory-core-host-engine-storage`, filed as openclaw/openclaw#134830. Per `docs/UPSTREAM-ADAPTATION-CONSTRAINT.md` we consume openclaw rather than change it, so the half we own is detect-and-notify — `scripts/verify_openclaw_memory_index.sh`, registered in **both** the `update` profile (a version bump forces a migration reindex, which is when this appeared) and the `runtime` profile (it recurs on openclaw's own schedule, not only after updates). Verified red against the live leak before commit.

### Also: `e3e4225` — the hook gate was reading the wrong log

`verify_openclaw_hooks.sh` resolved `$OPENCLAW_HOME/logs/gateway.log`, which exists on this machine but has been frozen since 2026-05-28 — the service writes to `~/Library/Logs/openclaw/`. The gate was reading a three-month-old file and reporting "no runtime signal": blind in exactly the way it exists to prevent. Now picks the newest known candidate and prints which log it read.

Worth noting for anyone debugging this class: openclaw hardcodes `LAUNCH_AGENT_STDERR_PATH = "/dev/null"` (`daemon/launchd-service-files.ts`) and regenerates the LaunchAgent on every `gateway restart`, so the memory subsystem's error stream is unobservable on a default macOS service install until you redirect it by hand — and it reverts on the next restart.

`pnpm run ci` green.

**Review by Amyssjj — COMMENTED** [review]
**Verdict: APPROVE**

PR has 14 files (+1219/-39), core config/install logic touching openclaw compat detection, memory-search layout migration, and conversation-hook grants → read as Large/risky (checked out the branch, read every changed source file and its test diff, not just the unified diff).

### Findings (0 blocking, 0 suggestion, 0 nit)

Three claimed blockers, each verified against the actual code rather than trusting the PR body:

1. **Conversation-hook gate** (`packages/cli/src/openclaw-memory.ts:225-241`, `packages/runtimes/openclaw/templates/recall/index.mjs`) — `mergeHookGrants` only sets `allowConversationAccess` when the key is `undefined` (never overrides an explicit `false`), and only for `digital-me-recall` (`CONVERSATION_HOOK_PLUGIN_IDS`), matching the least-privilege rationale (`digital-me-brain` registers no hooks). The recall plugin's own boot self-check (`conversationAccessGranted` → `conversation_hooks=granted|BLOCKED`) makes a silently-dropped hook visible in the log for the first time, and the `assistant_ack` fingerprint is now *derived* from what the host actually granted instead of hardcoded — `installer.test.ts` explicitly asserts the old hardcoded string is gone and the computed form is present, which is exactly the right thing to pin.

2. **Memory-search namespace cutover** (`openclaw-memory.ts:80-162`) — the legacy↔namespaced move is lossless (all 6 sub-keys preserved), bidirectional (forward AND rollback, tested at `openclaw-memory.test.ts`'s "migrates BACK when the host is rolled back below the cutover"), and correctly treats an unknown host version as "don't guess" (`isAtLeastOpenclawVersion` returns `undefined`, never coerced to `false` — confirmed by the dedicated test at `compat.test.ts:150-158`). The calendar-version comparator is verified against the actual gotcha it exists for: `isAtLeastOpenclawVersion("2026.10.1", "2026.9.1")` correctly returns `true` despite losing string comparison.

3. **Detection blind spots** — `scripts/verify_openclaw_hooks.sh` and `scripts/verify_openclaw_memory_index.sh` are read-only, pass config values via `argv` to the embedded Python (not string-interpolated — no injection risk), and both correctly resolve the *actual* running gateway log by newest-mtime rather than a hardcoded first-match path (the PR's own second commit found and fixed a case where the original log-path assumption was already stale — good catch caught by using the tool against a live system before merging).

Test quality: assertions pin actual behavior (config shape after migration, which key is absent, idempotency via byte-identical file snapshots) rather than "no exception thrown" or mock-call-count checks. Edge cases are deliberately covered: JSON5 dialect tolerance, non-object/malformed config, unknown host version, explicit `false` opt-out preserved, foreign `{path,pattern}` extraPaths entries preserved.

Scope: the second commit (reindex-leak detection) is outside the PR title's "survive the host cutover" framing, but it's disclosed and justified in a PR comment — found while live-verifying this same upgrade, same failure class (silent, all other signals green), and it lands as its own script + profile wiring rather than touching the cutover logic. Not scope creep in the agent-drift sense; it's a second real bug from the same verification pass.

CI is green across all 7 required checks (build/typecheck/lint/knip/test, digest, dream-cycle, sanitize gate) on this exact head commit.

### Summary
Clean, well-verified, well-tested change. No logic bugs found in the layout-migration state machine, the hook-grant merge, or the version-comparison boundary cases — all the failure modes the PR body claims to fix are backed by tests that would catch a regression, and the scripts added to close the detection gap are read-only and correctly scoped. Nothing to block on.



---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/verify_openclaw_hooks.sh [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  esac
+done
+
+# ── 2. Runtime: did the host or the plugin report a block? ──────────────
+# Only the tail matters — earlier boots may predate the fix.
+if [[ -n "$GATEWAY_LOG" && -f "$GATEWAY_LOG" ]]; then
+  note "log     $GATEWAY_LOG"
+  tail_txt=$(tail -c 2000000 "$GATEWAY_LOG" 2>/dev/null)
+
+  if grep -q "blocked because non-bundled plugins must set" <<<"$tail_txt"; then
```
Comment (Amyssjj) [inline scripts/verify_openclaw_hooks.sh]: Need fix: this greps **any** `blocked because non-bundled…` match in the last 2 MB of the log, not the last boot. An old `agent_end` gate warn (already dark on 6.11) can false-red the update sweep after the grant is written, until that line ages out of 2 MB.

The `conversation_hooks=` check below already uses last-registration-line-only. Mirror that for the host diagnostic — scope to the last boot / last `digital-me-recall: registered hooks` line — instead of the whole tail.

File: scripts/verify_openclaw_hooks.sh [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+fi
+
+for id in "${PLUGINS[@]}"; do
+  granted=$(python3 - "$CONFIG" "$id" <<'PY'
+import json, sys
+try:
+    # openclaw parses its config as JSON5; strict=False tolerates the control
+    # characters a hand-edited file can carry. A parse failure is reported as
+    # "unknown" rather than "ungranted" so this gate never fails on syntax.
+    cfg = json.loads(open(sys.argv[1]).read(), strict=False)
```
Comment (Amyssjj) [inline scripts/verify_openclaw_hooks.sh]: Need fix: `json.loads(..., strict=False)` is still JSON, not JSON5. OpenClaw accepts comments/trailing commas. An unparseable JSON5 config becomes `unknown` (warn, not fail), so a missing grant on a hand-annotated `openclaw.json` is a config-signal false green.

Parse with JSON5 (or the same dialect the installer uses) and fail when the grant is absent.

File: packages/runtimes/openclaw/templates/recall/index.mjs [hunk 3]
```diff
@@ -1060,13 +1100,36 @@ export default definePluginEntry({
...
+    );
     api.logger.info(
       `digital-me-recall: registered hooks (` +
         `boot=on, recall=${recallMaxResults > 0 ? "on" : "off"}, ` +
         `route=${enableRouteHashmap ? "on" : "off"}, ` +
         `observability=${enableObservability && brainDb ? "on" : "off"}, ` +
         `m1_emitter=${brainDb ? "on" : "wal-only"}, ` +
-        `assistant_ack=agent_end+before_message_write, ` +
+        `conversation_hooks=${convGranted ? "granted" : "BLOCKED"}, ` +
+        `assistant_ack=${ackSources.join("+") || "none"}, ` +
```
Comment (Amyssjj) [inline packages/runtimes/openclaw/templates/recall/index.mjs:1132]: Need fix: this log line is now a **computed** `assistant_ack=${ackSources.join("+")}`. Deploy (`digital-me.ts`) still greps `assistant_ack=([^,"`)\s]+)` out of the source/bundle, which captures the template-expression prefix, not the live value — so the fingerprint DIVERGE-false-reds or, if the regex misses, passes on any live marker (blind again).

The comment above says the fingerprint should reflect reality. Either emit a stable literal the grep can pin, or stop grepping source and match the live log against the derived set. Also: deploy still hardcodes `~/Library/Logs/openclaw/gateway.log` and does not share `resolve_gateway_log` from `verify_openclaw_hooks.sh`.

File: packages/transport/brain-mcp-proxy/src/handler.ts [hunk 4]
```diff
@@ -329,7 +331,7 @@ export function createCallToolHandler(deps: {
     };
 
     try {
-      const result = await invokeFn({ toolName: req.name, args });
+      const result = await invokeFn({ toolName: req.name, args, agentId });
```
Comment (Amyssjj) [inline packages/transport/brain-mcp-proxy/src/handler.ts]: Need fix: `agentId` here is `attributionLabel(args)` (the calling client, e.g. `hermes` / `codex`). Sending that as the OpenClaw 8.1 envelope owner is exactly the `Unknown agent id "hermes"` failure.

`server.ts` / `http-server.ts` overwrite this by wrapping `invokeFn` with `resolveGatewayAgentId`. A third caller of `createCallToolHandler` that trusts this field would still send attribution. Pass the gateway owner in, or drop `agentId` from this call and keep it only in the wrapper.

File: packages/transport/brain-mcp-proxy/src/config.ts [hunk 5]
```diff
@@ -136,6 +136,39 @@ export function loadGatewayConfig(input: {
...
+ *  - This one is the openclaw AGENT that owns the call. It must name a real
+ *    entry in `agents.entries`, because openclaw >= 2026.8.1 rejects an
+ *    invocation on a multi-agent host with no owner — and rejects an owner it
+ *    does not recognise just as hard (`Unknown agent id "hermes"`).
+ *
+ * Override with `$OPENCLAW_GATEWAY_AGENT_ID` on a host whose primary agent is
+ * not "main".
+ */
+export function resolveGatewayAgentId(env: NodeJS.ProcessEnv): string {
+  const raw = env.OPENCLAW_GATEWAY_AGENT_ID;
```
Comment (Amyssjj) [inline packages/transport/brain-mcp-proxy/src/config.ts:164]: Need fix: two env names for the same gateway **owner**.

- MCP proxy / dream-cycle / Hermes plugin / `m1_backfill.py` → `OPENCLAW_GATEWAY_AGENT_ID` (this file)
- Claude Code + Codex inject hooks and both `dm_m1_emit.py` → `DIGITAL_ME_OPENCLAW_AGENT_ID`

A host whose primary agent is not `main` (e.g. `coo`) can fix one path and leave the other rejecting. Pick `OPENCLAW_GATEWAY_AGENT_ID` and alias the hook var.

File: packages/runtimes/codex/hooks/dm_memory_search_inject.sh [hunk 6]
```diff
@@ -52,10 +52,18 @@ fi
...
+# \"main\" has no explicit owner. Pass agentId or use an agent-prefixed session
+# key." Without it every injection fails and the hook exits 0 silently, so
+# recall just stops with no error anywhere. Override per machine if the
+# default agent is not "main".
+AGENT_ID="${DIGITAL_ME_OPENCLAW_AGENT_ID:-main}"
+REQ="$(jq -cn --arg q "$QUERY" --arg a "$AGENT_ID" '{tool:"memory_search", agentId:$a, args:{query:$q, limit:6, corpus:"all"}}' 2>/dev/null)"
 [ -z "$REQ" ] && exit 0
 
-RESP="$(curl -sS -m 4 -X POST http://localhost:18789/tools/invoke \
+HOOK_TIMEOUT="${DIGITAL_ME_HOOK_TIMEOUT_SECS:-12}"
```
Comment (Amyssjj) [inline packages/runtimes/codex/hooks/dm_memory_search_inject.sh:66]: Need fix: curl default is 12s (`DIGITAL_ME_HOOK_TIMEOUT_SECS:-12`), but `buildCodexHooksManifest` sets this command’s `timeout` to **8** in `hooks.json`. Codex kills the hook before curl finishes, so a slow `memory_search` looks like a silent empty inject (the script `exit 0`s on empty RESP).

Make them the same number (raise the Codex timeout, or lower curl, and document which). Claude Code’s matching hook has the same 12s curl default — check its `hooks.json` timeout too.

File: packages/cli/src/doctor.ts [hunk 7]
```diff
@@ -472,19 +522,21 @@ export function runMemoryIndexChecks(deps: DoctorDeps): CheckResult[] {
...
-        `agents.defaults.memorySearch.extraPaths is missing ` +
+        `${msKey}.extraPaths is missing ` +
         `${missingPaths.join(", ")} — run 'digital-me install --runtime openclaw' ` +
         `(idempotent), then restart the openclaw gateway.`,
     });
   }
 
+  checks.push(runConversationHookGrantCheck(cfg, configPath));
+
   checks.push(runEmbeddingsCheck(deps.env, ms));
```
Comment (Amyssjj) [inline packages/cli/src/doctor.ts]: Need fix: extraPaths above is layout-aware (`memory.search` vs `agents.defaults.memorySearch`). `runEmbeddingsCheck` failure text is not — it still names `$OPENAI_API_KEY / agents.defaults.memorySearch.remote.apiKey` (around line 597) even when this host is on `memory.search`.

Pass `msKey` into the embeddings check so the repair hint points at a key that exists.

**Review by Amyssjj — COMMENTED** [review]
Maintainer review of leftover nits on this head (CI green). Overlay-prep (hook grant + memory layout migrate) looks correct. These are the items that still need a follow-up fix — not merge-blockers for the cutover default path (`agentId=main`).

Also not on a changed line, so called out here:

- `packages/cli/src/bin/digital-me.ts` deploy fingerprint still greps `assistant_ack=` out of the deployed source. After this PR that value is a template expression (`ackSources.join`), so the fingerprint either false-reds or goes blind. Compare the live log marker to the derived set, or drop the source grep and rely on `verify_openclaw_hooks.sh`.
- `mergeCodexMd` still uses the first `SECTION_END`. Hermes `mergeSoulMd` was fixed to `lastIndexOf` after a live 2-begin/9-end SOUL.md. Same accumulation is unhealed on CODEX.md.
- Installer user-facing copy in `digital-me.ts` still tells the operator to edit `agents.defaults.memorySearch` even on an 8.1 host.

No `@claude`. Merge stays yours.

**Comment by Amyssjj** [comment]
Addressed all three review findings in `5360546`. Each was verified against the code first; all three were real, and two were mine.

**1. Deploy fingerprint** — confirmed the false-red exactly as described. On the live host the source grep now yields `"${ackSources.join("` while the log reads `agent_end+before_message_write`, so every deploy would have diverged. Replaced with `expectedRecallAckMode(cfg)` in `deploy.ts`: derives the marker from `plugins.entries.digital-me-recall.hooks.allowConversationAccess`, the same grant the plugin reads. Verified derived == live on this host. Side benefit: a missing grant is now a deploy failure rather than a silent recall outage. 4 tests, including one asserting `parseRecallAckMode` and `expectedRecallAckMode` speak the same vocabulary.

**2. `mergeCodexMd`** — same two-part fix as hermes (strip embedded markers from the incoming section; span to `lastIndexOf(SECTION_END)`). The codex template also embeds its own markers at lines 11/132, same root cause. Live `CODEX.md` was 2 BEGIN / 9 END; reinstall now self-heals it to 1/1 with `config.toml` still parsing. 4 tests mirroring the hermes ones. I fixed one twin and missed the other — noted in memory as "grep for the structural twin before committing the first fix."

**3. Operator copy** — install and doctor messages are layout-aware now (`memory.search` on ≥2026.7.1, `agents.defaults.memorySearch` before). `runEmbeddingsCheck` takes the live key for its repair hint. Agreed it wasn't cosmetic: the legacy key on an 8.1 host hits a strict schema and the gateway refuses to start.

Node CI green; dream-cycle 171 / digest 50 against worktree source (reproduced with `PYTHONPATH=…/src`, since the venv otherwise resolves the main checkout).

**Comment by Amyssjj** [comment]
Two follow-up commits from today's live incident (COO `memory_search timed out after 15s`, 19/19 hourly heartbeats since the 8.1 cutover):

- **a34eafa** — brain-mcp-proxy: optional `agent` argument on `memory_search` / `memory_get` = the openclaw agent that owns the call (sent as the gateway's top-level `agentId`, overriding the process default). Needed because 8.1 gives one-shot CLI runs a *transient* memory manager (memory-core `tools.ts:318`, `oneShotCliRun → purpose "cli"`; the `memory.search` RPC does the same and closes it in a `finally`): a cold open of the COO's 1.7 GB store plus the new KNN subprocess cannot meet the hardcoded 15 s deadline, while the gateway's warm manager answers in ~1.3 s. 246 tests, coverage 100 %, verified against the live 8.1 host.
- **4b5194a** — both openclaw gates now read openclaw's debug file log (`/tmp/openclaw/openclaw-<date>.log`, what `openclaw logs` tails). The stdout log has no memory lines, so the memory-index gate was blind to 24 full-reindex aborts that stranded 8.9 GB of temp DBs today.

Still open on the host (not in this PR): each cold call also starts a full reindex that races the gateway's own indexer and aborts (upstream #134337 / #134332 / #134830); orphaned embedding-cache rows (878 MB) are never pruned. PR stays open for soak.

**Comment by Amyssjj** [comment]
Follow-up after live-testing the heartbeat path: the proxy `agent` route (a34eafa) does not help the COO heartbeat, because in a headless CLI-backend run openclaw treats an external MCP server as "native tool use" and the approval prompt can never be answered (`OpenClaw approval was not granted for native tool use (mcp__openclaw-brain__memory_search)`; standing grants are only minted for cron *exec* approvals).

- **bda6a6d** — `brain_memory_search {query, agent?, maxResults?}`: an in-gateway tool registered by the digital-me-brain plugin (`getActiveMemorySearchManager` from the plugin SDK = the gateway's warm manager). Plugin tools are served through the bundle server with no approval gate. 11 tests, package coverage 100 %.
- **683fda1** — declares it in the manifest `contracts.tools`; openclaw only serves plugin tools named there (the gateway logged "registered 8 tools" while `/tools/invoke` still said `Tool not available`).

Live: `/tools/invoke brain_memory_search agent=coo` → 6 hits in 0.9 s; a manual heartbeat run (`openclaw cron run …`) called `mcp__openclaw__brain_memory_search {query:"COO", agent:"coo"}` → count=6 → `HEARTBEAT_OK`, the first success since the 8.1 cutover. Root `pnpm run ci` green.

Operational note for anyone deploying this branch: `digital-me deploy` resets `~/digital-me-os` to **origin/main** before building, which silently reverts the soak; deploy the PR tip with `materializeOpenclawOverlay` from that checkout's built CLI plus `launchctl kickstart -k gui/$UID/ai.openclaw.gateway`.

**Review by Amyssjj — COMMENTED** [review]
**Verdict: APPROVE**

PR grew to 50 files / +2969/-96 / 16 commits since the last review (`59e2a478`, which covered the first 14-file cutover). Re-reviewing the incremental diff (`59e2a478..683fda1`, 19 files / +652/-31) rather than re-reading the whole branch, since the earlier review already went end-to-end on the base change and nothing there was touched again.

### Findings (0 blocking, 0 suggestion, 0 nit)

**Both prior blockers verified fixed, not just claimed:**
1. `packages/cli/src/deploy.ts:82-105` (`expectedRecallAckMode`) — replaces the old regex-over-deployed-source fingerprint (which broke the moment the marker became a template expression) with a derivation from the same `plugins.entries.digital-me-recall.hooks.allowConversationAccess` grant the plugin itself reads. `deploy.test.ts` pins the vocabulary match against `parseRecallAckMode`'s real log-line format and covers the `false`-not-`undefined` case.
2. `packages/runtimes/codex/src/installer.ts:78-95` (`mergeCodexMd`) — same two-part fix as the earlier hermes `SOUL.md` fix (strip embedded markers from the incoming section, span to `lastIndexOf(SECTION_END)`), applied to the twin bug on `CODEX.md`. `installer.test.ts`'s new suite constructs the exact damaged 2-BEGIN/9-END shape and asserts collapse, non-nesting, idempotency, and content preservation outside the section.

**New commits (`a34eafa`, `4b5194a`, `bda6a6d`, `683fda1`) — live-incident follow-up, in scope:**
- `brain-mcp-proxy`'s new `agent`/`ownerAgentId` param (`handler.ts:194-224`) is correctly kept distinct from `agent_id` (the attribution label) — `splitOwnerAgent` runs after `attributionLabel`, so the caller-identity log line is unaffected, and both `server.ts`/`http-server.ts` wire `ownerAgentId` ahead of the process default with `??`, not silently overriding always. Tests pin both the split (immutability, empty/non-string ignored-but-stripped) and the handler's full invoke-arg shape.
- `brain_memory_search` (`memory-search-tool.ts`) reuses the existing `OpenClawAgentTool` shape, is free of openclaw imports (search is injected), clamps `maxResults` into `[1, 20]`, and drops hits with no usable `path`. Registered in both `index.ts`'s exports and — the actual fix, not just the tool existing — `openclaw.plugin.json`'s `contracts.tools`, which is exactly the class of bug this same PR called out in its own body (`openclaw plugin only serves tools named there`).
- Both verify scripts' `resolve_log`/`resolve_gateway_log` additions (`/tmp/openclaw/openclaw-*.log` via `ls -t | head -1`) are read-only globs, safely handled when empty (`[[ -n "$cand" && -f "$cand" ]]`), consistent with the pattern already used for the other log-path candidates in the same functions.

CI is green on this exact head (`683fda1`): build/typecheck/lint/knip/test, dream-cycle, digest, sanitize gate all passing.

### Summary
Both previously-flagged review findings are fixed with real tests, not just described as fixed. The four new commits are a disclosed, in-scope continuation of the same live-verification pass (COO heartbeat timeouts after the 8.1 cutover) rather than unrelated scope creep, and the new `agent`-override / `brain_memory_search` code is well-isolated and tested at its boundaries (empty/non-string agent, missing path, over-limit results, no-owner-set). Nothing to block on.



---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Amyssjj — COMMENTED** [review]
**Verdict: REQUEST_CHANGES**

PR grew from 683fda1 (last reviewed) to 5e98eb9 with two more commits (8a9d7a7, 5e98eb9) addressing prior maintainer nits. Re-reviewing just the incremental diff (9 files, +166/-29): the JSON5-parse fix, last-boot log scoping, `resolveGatewayLog` extraction, env-var canonicalization, and hook timeout bump.

### Findings (2 blocking, 1 suggestion, 0 nit)

#### Blocking
1. `scripts/verify_openclaw_hooks.sh:78-107` — the new JSON5 parser fallback chain can't actually resolve on the host that runs it
   `health-sweep/profiles/update.json` invokes this script as `bash scripts/verify_openclaw_hooks.sh` from the **repo root** (`--repo .`). The Node branch does `require('json5')` inside a `node -` stdin script — Node resolves that via `process.cwd()`-based `node_modules` lookup, walking up from cwd, never into subdirectories. `json5` is declared only in `packages/cli/package.json` (`packages/cli/package.json:39`), not in the root `package.json`, and `.npmrc` sets `shamefully-hoist=false`, so pnpm will not hoist it to root `node_modules`. From repo root, `require('json5')` fails, `granted` stays empty (correctly triggering the fallback per the `|| granted=""` on the command substitution), but the Python fallback needs `pyjson5`, which is not declared anywhere in the repo (no `requirements*.txt`, no `pyproject.toml` reference — confirmed by grepping the whole tree). With neither parser resolvable, the script hits `raise SystemExit(1)` with "Install pyjson5" and hard-fails every run. This is the update-profile gate whose entire stated purpose is catching silent breakage that "nothing else can see" — it would now unconditionally fail (or require an undocumented manual `pip install pyjson5` per host), which is exactly the invisible-to-CI failure mode this whole PR is about. It also isn't caught by any GitHub Actions check — `health-sweep` pipelines aren't part of `ci.yml`.
   **Fix:** either resolve the npm `json5` package by explicit path (e.g. `NODE_PATH="$(dirname "$0")/../packages/cli/node_modules" node ...`, or `require(path.join(__dirname, ...))`) instead of relying on cwd-based resolution, or drop the external-package dependency entirely and hand-write a small JSON5-comment/trailing-comma stripper that's careful not to eat `//` inside quoted strings (the bug the old regex had) — that keeps the gate self-contained instead of depending on an unpinned, unverified runtime dependency.

2. `packages/cli/src/bin/digital-me.ts:2362-2367` — verification is now skipped (not retried) when the gateway log doesn't exist yet at restart time
   Before this diff, `logPath` was always a fixed, non-null path, and the following `while (Date.now() < deadline) { if (existsSync(logPath)) {...}; await sleepMs(2000); }` loop polled up to 30s for the file to *appear* — the right behavior right after `openclaw gateway restart`, since the daemon may not have written its log yet. Now `resolveGatewayLog` is called once, before the loop, and if none of its candidates exist yet it returns `null`, which takes the early `return true` — skipping the retry loop entirely and just assuming success. A first-boot host (or any host with zero existing gateway log anywhere: no `/tmp/openclaw/openclaw-*.log`, no `~/Library/Logs/openclaw/gateway.log`, no `$OPENCLAW_HOME/logs`, no XDG path) that then creates its log seconds later within the old 30s window would previously have caught a real registration failure and now silently reports "restart succeeded" without ever checking again.
   **Fix:** move the `resolveGatewayLog` call (or at least the null-check re-resolution) inside the polling loop so it keeps trying to find the log for the full 30s, the same way `existsSync(logPath)` used to keep trying to find content in it.

#### Suggestion
1. `packages/cli/src/deploy.ts:21-69` — `resolveGatewayLog`/`findNewestTmpLog` ship with zero tests
   `packages/cli/vitest.config.ts` sets 100% line/branch/function/statement thresholds over `src/**/*.ts` and explicitly excludes only `src/index.ts` and `src/bin/digital-me.ts` — `deploy.ts` is not excluded. `deploy.test.ts` is unchanged in this increment (`git diff 683fda1..5e98eb9 -- packages/cli/src/deploy.test.ts` is empty), so both new functions (env override, multi-candidate mtime comparison, empty-tmp-dir, unreadable-file catch branches) are currently uncovered — a real gap given every other function added across this PR's 18 commits got dedicated boundary tests. Worth noting the required GitHub Actions check runs `pnpm test`, not `pnpm test:coverage` (root `ci` script), so this gap won't fail CI even though it violates the package's own stated threshold — only `pnpm run ci` run locally would catch it.
   **Fix:** add a `resolveGatewayLog` test covering the env override, the multi-candidate newest-wins case, and the no-candidates → `null` path (which is exactly what finding #2 above depends on).

### Summary
@Amyssjj Two real regressions in this increment: the new JSON5 fallback chain in `verify_openclaw_hooks.sh` depends on a package (`json5`) that won't resolve from the script's actual invocation directory and a Python package (`pyjson5`) that's declared nowhere in the repo, so the gate this PR is adding will likely hard-fail on the very host it's meant to protect — worth actually running `bash scripts/verify_openclaw_hooks.sh` from repo root (not from `packages/cli`) to confirm. Separately, `resolveGatewayLog` being resolved once before the retry loop in `digital-me.ts` drops the old poll-for-log-to-appear behavior for a fresh/first-boot host. Both are easy, small fixes; the coverage gap in `deploy.ts` is a lower-priority follow-up.




---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Amyssjj** [comment]
Brought the branch up to date with `main` (merge of afeb6c7, the claude-code-action bump; no conflicts — GitHub's "out of date" prompt was the only thing showing) and restored the full local gate.

- **44c5b32** — merge `origin/main`.
- **c3def14** — tests for `resolveGatewayLog` / `findNewestTmpLog` (added in 8a9d7a7 without tests; the remote workflow runs `pnpm test` without thresholds, so it stayed green while `pnpm run ci` failed `@digital-me/cli` at 97 % lines). The tmp-log dir and fs calls are now injectable so the failure branches are testable without touching `/tmp/openclaw`. Package is back to 100 % on all four metrics; root `pnpm run ci` is green.

Host status after a week of soak on 2026.8.1: heartbeat green hourly via `brain_memory_search` (one retry absorbs the post-wake rebuild race, which is upstream behaviour and documented in the wiki). Ready to merge whenever you are.
